### PR TITLE
Optimise nested sessions code generation

### PR DESCRIPTION
**Pull request by becharrens**

### CODE REVIEW

File: lib/efsm.ml [hunk 1]
```diff
@@ -178,9 +187,41 @@ let powerset_construction (start, old_g) =
         Map.fold out_edges ~init:(g, state_map) ~f
   in
   let start_states = Map.find_exn epsilons start in
+  Stdio.print_endline
```
Comment (fangyi-zhou) [inline lib/efsm.ml:190]: Please remove this print.

Comment (becharrens) [reply lib/efsm.ml:190]: Sure, I was planning on undoing all the changes I made and keeping the latest version on develop. In the end I didn't use any of the changes I made to the efsm

File: lib/gocodegen.ml [hunk 2]
```diff
@@ -91,67 +91,104 @@ let ensure_unique_identifiers (global_t : global_t) =
...
   List.map ~f:extract_label ltypes
 
+let extract_choice_label_enums msgs_env ltypes =
+  let rec extract_enum_label = function
+    | AcceptL (_, protocol, roles, _, _, _) ->
+        MessagesEnv.get_invitation_enum msgs_env protocol roles
+    | RecvL ({label; _}, _, _) -> MessagesEnv.get_message_enum msgs_env label
+    | _ ->
+        raise
+          (Err.Violation
```
Comment (fangyi-zhou) [inline lib/gocodegen.ml:174]: There is a new function in `Err` called `violation` which basically raises a `Violation` (saves typing `raise`),
and another functional called `violationf` which is `violation` with `sprintf` for your convenience.

File: lib/lib.ml [hunk 3]
```diff
@@ -98,7 +95,7 @@ let enumerate_protocols (ast : scr_module) :
 
 let enumerate_nested_protocols (ast : scr_module) :
     (ProtocolName.t * RoleName.t) list =
-  let global_t = Gtype.global_t_of_ast ast in
+  let global_t = Gtype.global_t_of_module ast in
```
Comment (fangyi-zhou) [inline lib/lib.ml:98]: Closes #121 ?

**Pull request merged**